packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/string_prop.rs

```diff
@@ -17,52 +17,103 @@ pub struct StringProp {
+    ///   will always be `false` if there are any matching children.
+    ///
+    /// Regardless of the `propagate_came_from_default` argument,
+    /// - If there are two or more matching children, `came_from_default` is always `false`.
+    /// - If there are no matching children, then `came_from_default` will initially be true
+    ///   (and the value of the prop itself will be set to `default_value`).
+    ///   As soon as the value is changed (by a call to `invert()`),
+    ///   then `came_from_default` will be set to `false`.
+    pub fn new_from_children<S: Into<String>>(
+        default_value: S,
+        propagate_came_from_default: bool,
+    ) -> Self {
         StringProp {
             data_query: DataQuery::ChildPropProfile {
                 match_profiles: vec![ComponentProfile::String],
-                always_return_value: true,
+                always_return_value: propagate_came_from_default,
             },
-            default_value,
+            default_value: default_value.into(),
+            propagate_came_from_default,
         }
     }
 
     /// Creates a string prop that calculates its value from the attribute given by `attr_name`,
-    /// basing the calculation on the attribute children that match the `String` profile.
+    /// basing the calculation on the attribute components that match the `String` profile.
+    ///
+    /// If there are no matching attribute components, the prop will be initialized with `default_value`.
     ///
-    /// If there are no matching attribute children, the prop will be initialized with `default_value`.
-    pub fn new_from_attribute<S: Into<String>>(attr_name: AttributeName, default_value: S) -> Self {
+    /// The `propagate_came_from_default` argument influences
+    /// the behavior of this prop's `came_from_default` flag when there is only one matching attribute component
+    /// in the following manner:
+    /// - If `propagate_came_from_default` is `true`, then if there is only one matching attribute component,
+    ///   `came_from_default` will match the `came_from_default` of that single attribute component.
+    /// - If `propagate_came_from_default` is `false`, then `came_from_default`
+    ///   will always be `false` if there are any matching attribute components.
+    ///
+    /// Regardless of the `propagate_came_from_default` argument,
+    /// - If there are two or more matching attribute components, `came_from_default` is always `false`.
+    /// - If there are no matching attribute components, then `came_from_default` will initially be true
+    ///   (and the value of the prop itself will be set to `default_value`).
+    ///   As soon as the value is changed (by a call to `invert()`),
+    ///   then `came_from_default` will be set to `false`.
+    pub fn new_from_attribute<S: Into<String>>(
+        attr_name: AttributeName,
+        default_value: S,
+        propagate_came_from_default: bool,
```

Comment:
Why don't we have a consuming `.dont_propagate_came_from_default()` method here that returns a version that has `propagate_came_from_default` as `false`.

Reply:
I like this way of getting around Rust's refusal to provide default params!